nixpkgs/toolchains/cc.nix

```diff
@@ -176,10 +179,11 @@ in
           # Have gcc return the exit code from ld.
           add_compiler_option_if_supported -pass-exit-codes
         )
+      )
+      LINK_LIBS=(
         -lstdc++
         -lm
       )
-      LINK_LIBS=()
       OPT_COMPILE_FLAGS=(
```

Comment:
Without this we get linker errors of the form
```
bazel-out/k8-fastbuild/bin/tests/_objs/cc-test/cc-test.pic.o:tests/cc-test.cc:function main: error: undefined reference to 'std::cout'
bazel-out/k8-fastbuild/bin/tests/_objs/cc-test/cc-test.pic.o:tests/cc-test.cc:function main: error: undefined reference to 'std::basic_ostream<char, std::char_traits<char> >& std::operator<< <std::char_traits<char> >(std::basic_ostream<char, std::char_traits<char> >&, char const*)'
bazel-out/k8-fastbuild/bin/tests/_objs/cc-test/cc-test.pic.o:tests/cc-test.cc:function __cxx_global_var_init: error: undefined reference to 'std::ios_base::Init::Init()'
bazel-out/k8-fastbuild/bin/tests/_objs/cc-test/cc-test.pic.o:tests/cc-test.cc:function __cxx_global_var_init: error: undefined reference to 'std::ios_base::Init::~Init()'
clang-5.0: error: linker command failed with exit code 1 (use -v to see invocation)
```
It's unclear to me why this happens under clang. According to the output of `--verbose` gcc uses the same linker flags as clang does. The difference only lies in some `-L` flags and the use of `ld.gold` directly vs. `collect2` with `-fuse-ld=gold`.

Reply:
@aherrmann if you peak into nix's gcc/g++ there is a hack that checks for ++ in the filename: 

https://github.com/NixOS/nixpkgs/blob/8a78890a08dfa697be21891619f799738b956aa5/pkgs/build-support/cc-wrapper/cc-wrapper.sh#L28

Reply:
We're calling gcc as `cc` and clang as `clang`. So, both are invoked without the `++` in `argv[0]`. Passing `-x c++` on C++ compilation fixes the missing header issues for clang. What surprises me is that linking fails with clang on missing C++ symbols while it succeeds with gcc even though it looks like the relevant linker flags are identical.

gcc links as follows
```
/nix/store/vv4r320p5yd1k01kld62q1lppjxcswhb-gcc-7.3.0/libexec/gcc/x86_64-unknown-linux-gnu/7.3.0/collect2
  -plugin /nix/store/vv4r320p5yd1k01kld62q1lppjxcswhb-gcc-7.3.0/libexec/gcc/x86_64-unknown-linux-gnu/7.3.0/liblto_plugin.so
  -plugin-opt=/nix/store/vv4r320p5yd1k01kld62q1lppjxcswhb-gcc-7.3.0/libexec/gcc/x86_64-unknown-linux-gnu/7.3.0/lto-wrapper
  -plugin-opt=-fresolution=/run/user/1000/ccgR9Nbo.res
  -plugin-opt=-pass-through=-lgcc
  -plugin-opt=-pass-through=-lgcc_s
  -plugin-opt=-pass-through=-lc
  -plugin-opt=-pass-through=-lgcc
  -plugin-opt=-pass-through=-lgcc_s
  --eh-frame-hdr
  -m elf_x86_64
  -dynamic-linker /nix/store/fg4yq8i8wd08xg3fy58l6q73cjy8hjr2-glibc-2.27/lib64/ld-linux-x86-64.so.2
  -fuse-ld=gold
  -o bazel-out/k8-fastbuild/bin/tests/cc-test
  /nix/store/fg4yq8i8wd08xg3fy58l6q73cjy8hjr2-glibc-2.27/lib/crt1.o
  /nix/store/fg4yq8i8wd08xg3fy58l6q73cjy8hjr2-glibc-2.27/lib/crti.o
  /nix/store/vv4r320p5yd1k01kld62q1lppjxcswhb-gcc-7.3.0/lib/gcc/x86_64-unknown-linux-gnu/7.3.0/crtbegin.o
  -L/nix/store/fg4yq8i8wd08xg3fy58l6q73cjy8hjr2-glibc-2.27/lib
  -L/nix/store/zk5zj2307zxaq7dx585yia3dn5k4qlsl-gcc-7.3.0-lib/lib
  -L/nix/store/3xq3mgyrg968n5x58rhr95sgckhc1rhh-bazel-nixpkgs-cc/bin
  -L/nix/store/zk5zj2307zxaq7dx585yia3dn5k4qlsl-gcc-7.3.0-lib/lib
  -L/nix/store/fg4yq8i8wd08xg3fy58l6q73cjy8hjr2-glibc-2.27/lib
  -L/nix/store/iw94llkj05wgaz268mlzvgx8jkbi1ss0-gcc-wrapper-7.3.0/bin
  -L/nix/store/vv4r320p5yd1k01kld62q1lppjxcswhb-gcc-7.3.0/lib/gcc/x86_64-unknown-linux-gnu/7.3.0
  -L/nix/store/vv4r320p5yd1k01kld62q1lppjxcswhb-gcc-7.3.0/lib/gcc/x86_64-unknown-linux-gnu/7.3.0/../../../../lib64
  -L/nix/store/vv4r320p5yd1k01kld62q1lppjxcswhb-gcc-7.3.0/lib/gcc/x86_64-unknown-linux-gnu/7.3.0/../../..
  -dynamic-linker /nix/store/fg4yq8i8wd08xg3fy58l6q73cjy8hjr2-glibc-2.27/lib/ld-linux-x86-64.so.2
  -S
  -no-as-needed
  -z
  relro
  -z now
  -lstdc++
  -lm
  bazel-out/k8-fastbuild/bin/tests/_objs/cc-test/cc-test.pic.o
  -lgcc
  --as-needed -lgcc_s --no-as-needed
  -lc
  -lgcc
  --as-needed -lgcc_s --no-as-needed
  /nix/store/vv4r320p5yd1k01kld62q1lppjxcswhb-gcc-7.3.0/lib/gcc/x86_64-unknown-linux-gnu/7.3.0/crtend.o
  /nix/store/fg4yq8i8wd08xg3fy58l6q73cjy8hjr2-glibc-2.27/lib/crtn.o
```

clang links as follows
```
/nix/store/bsbbbfs80ywpwr0zp6bg3c0cj7s7n0bw-bazel-nixpkgs-cc/bin/ld.gold
  --eh-frame-hdr
  -m elf_x86_64
  -o bazel-out/k8-fastbuild/bin/tests/cc-test
  /nix/store/fg4yq8i8wd08xg3fy58l6q73cjy8hjr2-glibc-2.27/lib/crt1.o
  /nix/store/fg4yq8i8wd08xg3fy58l6q73cjy8hjr2-glibc-2.27/lib/crti.o
  /nix/store/vv4r320p5yd1k01kld62q1lppjxcswhb-gcc-7.3.0/lib64/gcc/x86_64-unknown-linux-gnu/7.3.0/crtbegin.o
  -L/nix/store/fg4yq8i8wd08xg3fy58l6q73cjy8hjr2-glibc-2.27/lib
  -L/nix/store/3qmlkci9qbflj03zg8c4f49gm2xjfkjd-clang-5.0.2-lib/lib
  -L/nix/store/vv4r320p5yd1k01kld62q1lppjxcswhb-gcc-7.3.0/lib64/gcc/x86_64-unknown-linux-gnu/7.3.0
  -L/nix/store/vv4r320p5yd1k01kld62q1lppjxcswhb-gcc-7.3.0/lib64/gcc/x86_64-unknown-linux-gnu/7.3.0/../../../../lib64
  -L/lib/x86_64-linux-gnu
  -L/lib/../lib64
  -L/usr/lib/x86_64-linux-gnu
  -L/usr/lib/../lib64
  -L/nix/store/vv4r320p5yd1k01kld62q1lppjxcswhb-gcc-7.3.0/lib64/gcc/x86_64-unknown-linux-gnu/7.3.0/../../..
  -L/nix/store/vknrja4x1j1ywih5vk5kizw1r5a0qpys-clang-5.0.2/bin/../lib
  -L/lib
  -L/usr/lib
  -dynamic-linker /nix/store/fg4yq8i8wd08xg3fy58l6q73cjy8hjr2-glibc-2.27/lib/ld-linux-x86-64.so.2
  -S
  -no-as-needed
  -z relro
  -z now
  -lstdc++
  -lm
  bazel-out/k8-fastbuild/bin/tests/_objs/cc-test/cc-test.pic.o
  -lgcc
  --as-needed -lgcc_s --no-as-needed
  -lc
  -lgcc
  --as-needed -lgcc_s --no-as-needed
  /nix/store/vv4r320p5yd1k01kld62q1lppjxcswhb-gcc-7.3.0/lib64/gcc/x86_64-unknown-linux-gnu/7.3.0/crtend.o
  /nix/store/fg4yq8i8wd08xg3fy58l6q73cjy8hjr2-glibc-2.27/lib/crtn.o
```

In both cases `-lstdc++` appears before `cc-test.pic.o`, so I'd expect missing symbols in both cases.

It seems that `collect2` somehow avoids these missing symbol errors. [Its description](https://gcc.gnu.org/onlinedocs/gccint/Collect2.html) lists special handling of C++ constructors. However, I don't find mention of libstdc++ symbols.